File: Python/network.py
import socket
import time
import os
import threading

ITERATIONS = 100
BUFFER_SIZE = 65536


def main():
    print("Python Network Benchmark")

    r, w = os.pipe()

    buffer = b"A" * BUFFER_SIZE

    t0 = time.perf_counter()
    for _ in range(ITERATIONS):
        os.write(w, buffer)
        data = os.read(r, BUFFER_SIZE)
    t1 = time.perf_counter()
    total_ms = (t1 - t0) * 1000

    throughput = (ITERATIONS * BUFFER_SIZE * 2) / (total_ms / 1000) / 1e9
    print(f"Pipe Throughput: {throughput:.2f} GB/s ({total_ms:.2f} ms)")

    import shutil

    data = bytearray(b"A" * 8192)
    dest = bytearray(8192)

    t0 = time.perf_counter()
    for _ in range(ITERATIONS * 10):
        dest[:] = data
    t1 = time.perf_counter()
    total_ms = (t1 - t0) * 1000

    throughput = (ITERATIONS * 10 * 8192) / (total_ms / 1000) / 1e9
    print(f"Memcpy 8K: {throughput:.2f} GB/s ({total_ms:.2f} ms)")

    os.close(r)
    os.close(w)


if __name__ == "__main__":
    main()

File: C++/network.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <unistd.h>
#include <string.h>

#define ITERATIONS 100
#define BUFFER_SIZE 65536

using namespace std;
using namespace chrono;

int main() {
    cout << "C++ Network Benchmark" << endl;
    
    int fds[2];
    pipe(fds);
    
    vector<char> buffer(BUFFER_SIZE, 'A');
    
    auto t0 = high_resolution_clock::now();
    for (int i = 0; i < ITERATIONS; i++) {
        write(fds[1], buffer.data(), BUFFER_SIZE);
        read(fds[0], buffer.data(), BUFFER_SIZE);
    }
    auto t1 = high_resolution_clock::now();
    double total_ms = duration<double, milli>(t1 - t0).count();
    
    double throughput = (ITERATIONS * BUFFER_SIZE * 2) / (total_ms / 1000.0) / 1e9;
    cout << "Pipe Throughput: " << throughput << " GB/s (" << total_ms << " ms)" << endl;
    
    vector<char> src(8192, 'A');
    vector<char> dst(8192, 0);
    
    t0 = high_resolution_clock::now();
    for (int i = 0; i < ITERATIONS * 10; i++) {
        memcpy(dst.data(), src.data(), 8192);
    }
    t1 = high_resolution_clock::now();
    total_ms = duration<double, milli>(t1 - t0).count();
    
    throughput = (ITERATIONS * 10 * 8192) / (total_ms / 1000.0) / 1e9;
    cout << "Memcpy 8K: " << throughput << " GB/s (" << total_ms << " ms)" << endl;
    
    close(fds[0]);
    close(fds[1]);
    
    return 0;
}
